tests/publisher.cpp

```diff
@@ -67,17 +69,103 @@ BOOST_AUTO_TEST_CASE(multiple_publisher_on_same_host)
         return;
 
-    const zeq::uint128_t uuid = servus::make_UUID();
-    const std::string scheme = "schema-" + std::to_string( uuid.high( )) +
-                               std::to_string( uuid.low( ));
-    const servus::URI uri( scheme + "://*:0" );
+    const zeq::Publisher publisher1;
+    const zeq::Publisher publisher2;
+    const zeq::Publisher publisher3;
 
-    const zeq::Publisher publisher1( uri );
-    const zeq::Publisher publisher2( uri );
-    const zeq::Publisher publisher3( uri );
-
-    servus::Servus service( std::string( "_" ) + scheme + "._tcp" );
+    servus::Servus service( PUBLISHER_SERVICE );
     const servus::Strings& instances =
             service.discover( servus::Servus::IF_ALL, 1000 );
     BOOST_CHECK_EQUAL( instances.size(), 3 );
 }
+
+BOOST_AUTO_TEST_CASE(zeroconf_record)
+{
+    if( !servus::Servus::isAvailable() || getenv("TRAVIS"))
+        return;
+
+    const zeq::Publisher publisher;
+
+    servus::Servus service( PUBLISHER_SERVICE );
+    const servus::Strings& instances =
+            service.discover( servus::Servus::IF_ALL, 1000 );
+    BOOST_REQUIRE_EQUAL( instances.size(), 1 );
+
+    const std::string& instance = instances[0];
+    BOOST_CHECK_EQUAL( instance, publisher.getAddress( ));
+    BOOST_CHECK_EQUAL( service.get( instance, KEY_APPLICATION ), "publisher" );
+    BOOST_CHECK_EQUAL( zeq::uint128_t( service.get( instance, KEY_INSTANCE )),
+                       zeq::detail::Sender::getUUID( ));
+    BOOST_CHECK_EQUAL( service.get( instance, KEY_SESSION ), getUserName( ));
+    BOOST_CHECK_EQUAL( service.get( instance, KEY_USER ), getUserName( ));
+}
+
+BOOST_AUTO_TEST_CASE(custom_session)
+{
+    if( !servus::Servus::isAvailable() || getenv("TRAVIS"))
+        return;
+
+    const zeq::Publisher publisher( test::buildUniqueSession( ));
+
+    servus::Servus service( PUBLISHER_SERVICE );
+    const servus::Strings& instances =
+            service.discover( servus::Servus::IF_ALL, 1000 );
+    BOOST_REQUIRE_EQUAL( instances.size(), 1 );
+
+    const std::string& instance = instances[0];
+    BOOST_CHECK_EQUAL( service.get( instance, KEY_SESSION ),
+                       publisher.getSession( ));
+}
+
+BOOST_AUTO_TEST_CASE(different_session_at_runtime)
+{
+    if( !servus::Servus::isAvailable() || getenv("TRAVIS"))
+        return;
+
+    setenv( "ZEROEQ_SESSION", "testsession", 1 );
+    const zeq::Publisher publisher;
+
+    servus::Servus service( PUBLISHER_SERVICE );
+    const servus::Strings& instances =
+            service.discover( servus::Servus::IF_ALL, 1000 );
+    BOOST_REQUIRE_EQUAL( instances.size(), 1 );
+
+    const std::string& instance = instances[0];
+    BOOST_CHECK_EQUAL( service.get( instance, KEY_SESSION ),
+                       "testsession" );
+    unsetenv(  "ZEROEQ_SESSION" );
+}
+
+BOOST_AUTO_TEST_CASE(empty_session)
+{
+    BOOST_CHECK_THROW( const zeq::Publisher publisher( "" ),
+                       std::runtime_error );
+}
+
```

Comment:
Defaulting to username if ZEROEQ_SESSION is empty is fine for me.

Regarding the session class I think it introduces a complication on the API unless it can be default constructed from std::string and cosnt char*. If you read the documentation I think it's pretty clear that Publisher( "" ) throws. Actually it's good that it throws because it doesn't leave the user wondering what would happen if accepted.
The only benefit I see about having the class is that you can query the default session before passing it to the Publisher if needed.